File: 101.对称二叉树.c
/*
 * @lc app=leetcode.cn id=101 lang=c
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool sym(struct TreeNode *left, struct TreeNode *right)
{
  // 未化简
  if (left == NULL && right != NULL)
    return false;
  else if (left != NULL && right == NULL)
    return false;
  else if (left == NULL && right == NULL)
    return true;
  else if (left->val != right->val)
    return false;
  bool in = sym(left->right, right->left);
  bool out = sym(left->left, right->right);
  return in && out;
}

bool isSymmetric(struct TreeNode *root)
{
  // 难点：不是比较节点大小，是比较子树
  // 法一：递归
  // if (root == NULL)
  //   return true;
  // return sym(root->left, root->right);

  // 法二：队列迭代
  // struct TreeNode **queue, *left, *right;
  // int front, rear;
  // queue = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 1000);
  // front = rear = 0;
  // if (root != NULL)
  // {
  //   queue[rear++] = root->left;
  //   queue[rear++] = root->right;
  //   while (front != rear)
  //   {
  //     left = queue[front++];
  //     right = queue[front++];
  //     // 指针后需要对其进行判空
  //     if (left == NULL && right == NULL)
  //       continue;

  //     if ((left == NULL && right != NULL) || (left != NULL && right == NULL) || (left->val != right->val))
  //       return false;

  //     queue[rear++] = left->right;
  //     queue[rear++] = right->left;
  //     queue[rear++] = left->left;
  //     queue[rear++] = right->right;
  //   }
  // }
  // return true;

  // 法三：栈迭代
  struct TreeNode **stack, *left, *right;
  int top;
  stack = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 1000);
  top = 0;
  if (root != NULL)
  {
    stack[top++] = root->left;
    stack[top++] = root->right;
    while (top > 0)
    {
      left = stack[--top];
      right = stack[--top];
      if (left == NULL && right == NULL)
        continue;
      if (left == NULL || right == NULL || (left->val != right->val))
        return false;
      stack[top++] = left->right;
      stack[top++] = right->left;
      stack[top++] = left->left;
      stack[top++] = right->right;
    }
  }
  return true;
}
// @lc code=end

File: 101.对称二叉树.js
/*
 * @lc app=leetcode.cn id=101 lang=javascript
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const Sym = (left, right) => {
  if (left && !right) return false;
  if (!left && right) return false;
  if (!left && !right) return true;
  if (left && right && left.val !== right.val) return false;
  if (left && right && left.val === right.val) {
    let leftSym = Sym(left.left, right.right);
    let rightSym = Sym(left.right, right.left);
    return leftSym && rightSym;
  }
};

var isSymmetric = function (root) {
  // 法一
  // return Sym(root.left, root.right);
  // 法二
  let queue = [],
    front,
    rear;
  front = rear = 0;
  queue[rear++] = root.left;
  queue[rear++] = root.right;
  while (front !== rear) {
    let left = queue[front++];
    let right = queue[front++];
    if (!left && right) return false;
    else if (left && !right) return false;
    else if (left && right && left.val !== right.val) return false;
    else if (left && right && left.val === right.val) {
      queue[rear++] = left.left;
      queue[rear++] = right.right;
      queue[rear++] = left.right;
      queue[rear++] = right.left;
    }
  }
  return true;
};
// @lc code=end
